packages/flutter/lib/src/widgets/tween_animation_builder.dart

```diff
@@ -212,3 +216,168 @@ class _TweenAnimationBuilderState<T extends Object?>
+
+  /// Called every time the animation value changes.
+  ///
+  /// The builder receives the current animation value and the optional child.
+  final ValueWidgetBuilder<T> builder;
+
+  /// The child widget to pass to the builder.
+  ///
+  /// If a builder callback's return value contains a subtree that does not
+  /// depend on the animation, it's more efficient to build that subtree once
+  /// instead of rebuilding it on every animation tick.
+  final Widget? child;
+
+  /// Whether the animation should reverse direction when it reaches the end.
+  ///
+  /// When true, the animation reverses direction instead of jumping back to
+  /// the beginning. Defaults to false.
+  final bool reverse;
+
+  /// Whether the animation is currently paused.
+  ///
+  /// When true, the animation stops at its current value. When changed to
+  /// false, the animation resumes from that value. Defaults to false.
+  final bool paused;
+
+  @override
+  State<RepeatingTweenAnimationBuilder<T>> createState() {
+    return _RepeatingTweenAnimationBuilderState<T>();
+  }
+}
+
+class _RepeatingTweenAnimationBuilderState<T extends Object>
+    extends State<RepeatingTweenAnimationBuilder<T>>
+    with SingleTickerProviderStateMixin {
+  late final AnimationController _controller;
+  late Animation<T> _animation;
+  CurvedAnimation? _curvedAnimation;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = AnimationController(duration: widget.duration, vsync: this);
+    _updateAnimation();
+
+    // Start the animation if not paused and begin != end.
+    if (!widget.paused && widget.tween.begin != widget.tween.end) {
+      _controller.repeat(reverse: widget.reverse);
+    }
+  }
+
+  void _updateAnimation() {
+    _curvedAnimation?.dispose();
+    _curvedAnimation = CurvedAnimation(parent: _controller, curve: widget.curve);
+    _animation = widget.tween.animate(_curvedAnimation!);
+  }
+
+  @override
+  void didUpdateWidget(RepeatingTweenAnimationBuilder<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+
+    // Update the duration if it changed.
+    if (widget.duration != oldWidget.duration) {
+      _controller.duration = widget.duration;
+    }
+
+    // Update the animation if the curve or the tween changed.
+    final bool tweenChanged =
+        widget.tween.runtimeType != oldWidget.tween.runtimeType ||
+        widget.tween.begin != oldWidget.tween.begin ||
+        widget.tween.end != oldWidget.tween.end;
+    if (widget.curve != oldWidget.curve || tweenChanged) {
+      _updateAnimation();
+    }
+
+    final bool shouldAnimate = widget.tween.begin != widget.tween.end;
+    final bool wasAnimating = oldWidget.tween.begin != oldWidget.tween.end;
+
+    if (!shouldAnimate || widget.paused) {
+      _controller.stop(canceled: false);
+    } else if (!wasAnimating ||
+        widget.paused != oldWidget.paused ||
+        widget.reverse != oldWidget.reverse) {
+      _controller.repeat(reverse: widget.reverse);
```

Comment:
Hm a slight drawback here is that this will always goes forward, even if we paused while the animation was running in reverse. Unfortunately, this seems to be a limitation of the `AnimationController` API (we might need to add `AnimationController.pause()` and `AnimationController.resume()` methods to address this?). I think that's OK for now, we can always fix this later.

Reply:
The AnimationController API is very VERY confusing, and I guess part of that went for our new Tween. Right now, "reverse" means: when it reaches the end it "reverses" until zero. Like 0->1->0, instead of 0->1, 0->1. It doesn't mean, like, "please revert the movement in the opposite direction"

At some point I considered cleaning/improving the AnimationController, but it would be breaking. Not sure if we should just (eventually) make InfiniteAnimationController or something like that, which Compose does (and is much cleaner).